rtgui/editcallbacks.h

```diff
@@ -0,0 +1,256 @@
+ *
+ *  Copyright (c) 2019 Jean-Christophe FRISCH <natureh.510@gmail.com>
+ *
+ *  RawTherapee is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  RawTherapee is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with RawTherapee.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include "editid.h"
+#include "cursormanager.h"
+#include "../rtengine/coord.h"
+
+class Geometry;
+class EditDataProvider;
+
+/** @file
+ * See editwidgets.h for documentation
+ */
+
+/// @brief Method for client tools needing Edit information
+class EditSubscriber
+{
+
+public:
+
+private:
+    EditUniqueID ID; /// this will be used in improcfun to locate the data that has to be stored in the buffer; it must be unique in RT
+    EditType editingType;
+    BufferType bufferType;
+    EditDataProvider *provider;
+
+protected:
+    std::vector<Geometry*> visibleGeometry;    /// displayed geometry
+    std::vector<Geometry*> mouseOverGeometry;  /// mouseOver geometry, drawn in a hidden buffer
+    enum {
+        ES_ACTION_NONE,      ///
+        ES_ACTION_DRAGGING,  /// set action to this value in the buttonPressed event to start dragging and ask for drag event
+        ES_ACTION_PICKING    /// set action to this value in the buttonPressed event whenever the user is picking something through a single click. In this case, the pickX events will be called INSTEAD of buttonXReleased !
+    } action;                /// object mode only, ignored in Pipette mode
+
+public:
+    explicit EditSubscriber (EditType editType);
+    virtual ~EditSubscriber () {}
+
+    void               setEditProvider(EditDataProvider *provider);
+    EditDataProvider*  getEditProvider ();
+    void               setEditID(EditUniqueID ID, BufferType buffType);
+    bool               isCurrentSubscriber();
+    virtual void       subscribe();
+    virtual void       unsubscribe();
+    virtual void       switchOffEditMode ();    /// Occurs when the user want to stop the editing mode
+    EditUniqueID       getEditID();
+    EditType           getEditingType();
+    BufferType         getPipetteBufferType();
+    bool               isDragging();            /// Returns true if something is being dragged and drag events has to be sent (object mode only)
+    bool               isPicking();             /// Returns true if something is being picked
+
+    /** @brief Get the cursor to be displayed when above handles
+    @param objectID object currently "hovered" */
+    virtual CursorShape getCursor (const int objectID);

```

Comment:
IMHO you should either inline the default behavior here directly or move the implementation to the `.cc`. Splitting it to two places in the header sometimes makes sense for template implementation but not so much for standard methods.

The `const int objectID` should really be `int objectID` as this is a top level `const` in a declaration. See my comment and the links [here](https://github.com/CarVac/librtprocess/issues/29#issuecomment-461724660).

Reply:
I don't know what `top level` means in this context, but I removed it.